C front-end: reject code containing duplicate labels

tautschnig opened the pull request:
We previously silently accepted such code, maintaining the labels and
attaching them to multiple instructions. `goto` resolved to the first
such occurrence of the label.

GCC firmly rejects such code.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
C front-end: reject code containing duplicate labels

We previously silently accepted such code, maintaining the labels and
attaching them to multiple instructions. `goto` resolved to the first
such occurrence of the label.

GCC firmly rejects such code.

NlightNFotis commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -505,7 +505,12 @@ void c_typecheck_baset::typecheck_for(codet &code)
+  if(!labels_defined.emplace(code.get_label(), code.source_location()).second)
+  {
+    error().source_location = code.source_location();
+    error() << "duplicate label '" << code.get_label() << "'" << eom;
```
💡 Can this be substituted for a `throw` of an `invalid_source_file_exceptiont` from `src/utils/exception_utils.h`? That would be great!

tautschnig replied:
I'm going to do that once #5835 and #5837 have been merged.

NlightNFotis reviewed (commented):
Solid, thanks for that!

tautschnig committed:
Ensure labels (even __CPROVER_ASYNC_*) are unique per function

An upcoming change will enforce per-function unique labels, as mandated
by the C standard. We were sometimes sloppy about this with
__CPROVER_ASYNC_* pseudo-labels as they don't actually indicate jump
targets.

tautschnig committed:
Clang-format code after updates to labels

Whitespace changes only.
